### PR TITLE
Refactoring of entire structure, new base image

### PR DESCRIPTION
**Pull request by tboerger**
* Added
  * Added scripts folder before and after cronjob
* Changed
  * Change to environment-based configuration
  * Reduce amount of shell script magic
  * Use the new `owncloud/php` Docker image
  * Combined `owncloud-*` scripts to `owncloud`
  * Moved templates to `/etc/templates`
  * Reduced amount of required templates
* Removed
  * Removed Apache configuration, part of base image now
  * Removed `/usr/local/bin` scripts, moved to `/usr/bin`

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1 +1,3 @@
+.envrc
```
Comment (patrickjahns) [inline .dockerignore]: Where is that from?

File: rootfs/usr/bin/server [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+if [[ -d "${OWNCLOUD_PRE_SERVER_PATH}" ]]
+then
+  for FILE in $(find ${OWNCLOUD_PRE_SERVER_PATH} -iname *.sh | sort)
+  do
+    source ${FILE}
+  done
+fi
+
+echo "Starting apache daemon..."
+exec apachectl -f /etc/apache2/apache2.conf -DFOREGROUND
```
Comment (patrickjahns) [inline rootfs/usr/bin/server]: isn't that blocking?

File: .dockerignore (hunk 1 above)
Comment (tboerger) [reply .dockerignore]: Used by https://github.com/direnv/direnv, just to always avoid the addition of that file.

File: rootfs/usr/bin/server (hunk 2 above)
Comment (tboerger) [reply rootfs/usr/bin/server]: yes, that's blocking and that should be fine because dockerize have been dropped.

**Comment by patrickjahns** [comment]
@tboerger 
Did you test the cronjob part?
https://github.com/owncloud-docker/base/pull/64/files#diff-d582bc91408e45b45431d50d32f4d8ff

I seem to remember that the normal environment vars are not present in the cron process

**Comment by tboerger** [comment]
> Did you test the cronjob part?
> https://github.com/owncloud-docker/base/pull/64/files#diff-d582bc91408e45b45431d50d32f4d8ff
> 
> I seem to remember that the normal environment vars are not present in the cron process

The cron process is not started via `service` anmore, that's why the variables seem to be present so far. Let me confirm that the job is correctly processed.

File: rootfs/etc/entrypoint.d/85-others.sh [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+[[ -z "${OWNCLOUD_UPGRADE_AUTOMATIC_APP_UPDATES}" ]] && OWNCLOUD_UPGRADE_AUTOMATIC_APP_UPDATES=""
+
+declare -x OWNCLOUD_FILES_EXTERNAL_ALLOW_NEW_LOCAL
+[[ -z "${OWNCLOUD_FILES_EXTERNAL_ALLOW_NEW_LOCAL}" ]] && OWNCLOUD_FILES_EXTERNAL_ALLOW_NEW_LOCAL="${OWNCLOUD_ALLOW_EXTERNAL_LOCAL_STORAGE}"
+
+declare -x OWNCLOUD_SMB_LOGGING_ENABLE
+[[ -z "${OWNCLOUD_SMB_LOGGING_ENABLE}" ]] && OWNCLOUD_SMB_LOGGING_ENABLE=""
+
+declare -x OWNCLOUD_DAV_ENABLE_ASYNC
+[[ -z "${OWNCLOUD_DAV_ENABLE_ASYNC}" ]] && OWNCLOUD_DAV_ENABLE_ASYNC=""
```
Comment (patrickjahns) [inline rootfs/etc/entrypoint.d/85-others.sh]: that's a tech-preview feature - not sure if we should expose this currently

File: rootfs/etc/entrypoint.d/55-marketplace.sh [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
+declare -x OWNCLOUD_MARKETPLACE_KEY
+[[ -z "${OWNCLOUD_MARKETPLACE_KEY}" ]] && OWNCLOUD_MARKETPLACE_KEY="${OWNCLOUD_MARKETPLACE_APIKEY}"
 
 declare -x OWNCLOUD_MARKETPLACE_CA
 [[ -z "${OWNCLOUD_MARKETPLACE_CA}" ]] && OWNCLOUD_MARKETPLACE_CA=""
 
-declare -x OWNCLOUD_MARKETPLACE_APIKEY
-[[ -z "${OWNCLOUD_MARKETPLACE_APIKEY}" ]] && OWNCLOUD_MARKETPLACE_APIKEY=""
+declare -x OWNCLOUD_APPSTORE_URL
+[[ -z "${OWNCLOUD_APPSTORE_URL}" ]] && OWNCLOUD_APPSTORE_URL="${OWNCLOUD_MARKETPLACE_URL}"
```
[the remark this answers is not in the document]
Comment (tboerger) [reply rootfs/etc/entrypoint.d/55-marketplace.sh]: It's the only variable to change the URL for the marketplace, and since it's a documented core config variable and so far the env variables match the core config name I have renamed it.

**Comment by patrickjahns** [comment]
How was this tested so far?
Are there any test scenarios where you upgrade from a previous image to this new image?

File: rootfs/etc/entrypoint.d/60-apps.sh [hunk 5]
```diff
@@ -6,4 +6,10 @@ declare -x OWNCLOUD_APPS_INSTALL
 declare -x OWNCLOUD_APPS_ENABLE
 [[ -z "${OWNCLOUD_APPS_ENABLE}" ]] && OWNCLOUD_APPS_ENABLE="${OWNCLOUD_APPS_INSTALL}"
 
+declare -x OWNCLOUD_APPS_UNINSTALL
```
[the remark this answers is not in the document]
Comment (tboerger) [reply rootfs/etc/entrypoint.d/60-apps.sh]: Uninstalling apps :)
So far I have done that via some custom shell script if needed.

File: rootfs/etc/entrypoint.d/85-others.sh (hunk 3 above)
Comment (tboerger) [reply rootfs/etc/entrypoint.d/85-others.sh]: But it's still part of the core sample config, I have added all simple variables, just some special array variables are dropped from the env variables.

**Comment by tboerger** [comment]
> How was this tested so far?
> Are there any test scenarios where you upgrade from a previous image to this new image?

So far only basic manual testing.

**Comment by patrickjahns** [comment]
@tboerger - where the `config.php` options taken from `master` or `stable10` - they might differ currently

**Comment by tboerger** [comment]
Indeed taken from master, but the diff is not that large:

```diff
# diff -ruN stable10-config.php master-config.php
--- stable10-config.php	2018-10-09 12:47:23.000000000 +0200
+++ master-config.php	2018-10-09 12:47:07.000000000 +0200
@@ -35,7 +35,7 @@
  * This is a unique identifier for your ownCloud installation, created
  * automatically by the installer. This example is for documentation only,
  * and you should never use it because it will not work. A valid ``instanceid``
- * is created when you install ownCloud.
+ * is created when you install ownCloud. Needs to start with a letter.
  *
  * 'instanceid' => 'd3c944a9a',
  */
@@ -415,9 +415,10 @@

 /**
  * This option allows you to define a manual override condition as a regular
- * expression for the remote IP address. For example, defining a range of IP
- * addresses starting with ``10.0.0.`` and ending with 1 to 3:
- * ``^10\.0\.0\.[1-3]$``
+ * expression for the remote IP address. The keys ``overwritewebroot``,
+ * ``overwriteprotocol``, and ``overwritehost`` are subject to this condition.
+ * For example, defining a range of IP  addresses starting with ``10.0.0.``
+ * and ending with 1 to 3: * ``^10\.0\.0\.[1-3]$``
  */
 'overwritecondaddr' => '',

@@ -492,9 +493,9 @@
  * Available values:
  *
  * * ``auto``
- *     default setting. keeps files and folders in the trash bin for 30 days
- *     and automatically deletes anytime after that if space is needed (note:
- *     files may not be deleted if space is not needed).
+ *     default setting. Keeps files and folders in the deleted files for up to
+ *     30 days, automatically deleting them (at any time) if space is needed.
+ *     Note: files may not be removed if space is not required.
  * * ``D, auto``
  *     keeps files and folders in the trash bin for D+ days, delete anytime if
  *     space needed (note: files may not be deleted if space is not needed)
@@ -561,13 +562,6 @@
  */

 /**
- * Checks an app before install whether it uses private APIs instead of the
- * proper public APIs. If this is set to true it will only allow to install or
- * enable apps that pass this check.
- */
-'appcodechecker' => true,
-
-/**
  * Check if ownCloud is up-to-date and shows a notification if a new version is
  * available. This option is only applicable to ownCloud core. It is not
  * applicable to app updates.
@@ -700,12 +694,6 @@
 'logtimezone' => 'Europe/Berlin',

 /**
- * Append all database queries and parameters to the log file. Use this only for
- * debugging, as your logfile will become huge.
- */
-'log_query' => false,
-
-/**
  * Log successful cron runs.
  */
 'cron_log' => true,
@@ -762,11 +750,6 @@
     ),

 /**
- * @see appcodechecker
- */
-
-
-/**
  * Previews
  *
  * ownCloud supports previews of image files, the covers of MP3 files, and text
@@ -1075,56 +1058,6 @@
 'dav.chunk_base_dir' => '',

 /**
- * Using Object Store with ownCloud
- */
-
-/**
- * This example shows how to configure ownCloud to store all files in a
- * swift object storage.
- *
- * It is important to note that ownCloud in object store mode will expect
- * exclusive access to the object store container because it only stores the
- * binary data for each file. The metadata is currently kept in the local
- * database for performance reasons.
- *
- * WARNING: The current implementation is incompatible with any app that uses
- * direct file IO and circumvents our virtual filesystem. That includes
- * Encryption and Gallery. Gallery will store thumbnails directly in the
- * filesystem and encryption will cause severe overhead because key files need
- * to be fetched in addition to any requested file.
- *
- * One way to test is applying for a trystack account at http://trystack.org/
- */
-'objectstore' => [
-	'class' => 'OC\\Files\\ObjectStore\\Swift',
-	'arguments' => [
-		// trystack will use your facebook id as the user name
-		'username' => 'facebook100000123456789',
-		// in the trystack dashboard go to user -> settings -> API Password to
-		// generate a password
-		'password' => 'Secr3tPaSSWoRdt7',
-		// must already exist in the objectstore, name can be different
-		'container' => 'owncloud',
-		// prefix to prepend to the fileid, default is 'oid:urn:'
-		'objectPrefix' => 'oid:urn:',
-		// create the container if it does not exist. default is false
-		'autocreate' => true,
-		// required, dev-/trystack defaults to 'RegionOne'
-		'region' => 'RegionOne',
-		// The Identity / Keystone endpoint
-		'url' => 'http://8.21.28.222:5000/v2.0',
-		// required on dev-/trystack
-		'tenantName' => 'facebook100000123456789',
-		// dev-/trystack uses swift by default, the lib defaults to 'cloudFiles'
-		// if omitted
-		'serviceName' => 'swift',
-		// The Interface / url Type, optional
-		'urlType' => 'internal'
-	],
-],
-
-
-/**
  * Sharing
  *
  * Global settings for Sharing
```

**Comment by tboerger** [comment]
You have been right, cron was missing the env variables. I'm writing the env to `/etc/environment` now:

```diff
diff --git a/rootfs/etc/entrypoint.d/98-overwrite.sh b/rootfs/etc/entrypoint.d/98-overwrite.sh
index 104d017..ec7b292 100755
--- a/rootfs/etc/entrypoint.d/98-overwrite.sh
+++ b/rootfs/etc/entrypoint.d/98-overwrite.sh
@@ -8,4 +8,7 @@ export APACHE_ACCESS_LOG="${OWNCLOUD_ACCESSLOG_LOCATION:-/dev/stderr}"
 export APACHE_DOCUMENT_ROOT="${APACHE_DOCUMENT_ROOT:-/var/www/owncloud}"
 export APACHE_LISTEN="${APACHE_LISTEN:-8080}"

+# Write env to file for crond
+env >| /etc/environment
+
 true
```

File: .dockerignore (hunk 1 above)
Comment (tboerger) [reply .dockerignore]: Dropped, added it to my global gitignore file since it's not relevant for this project.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,21 @@
 # Changelog
 
+## 2018-10-09
+
+* Added
+  * Added scripts folder before and after cronjob
+  * Test built image with Clair before publishing
+* Changed
```
[the remark this answers is not in the document]
Comment (tboerger) [reply CHANGELOG.md]: Added a `Deprecated`section to changelog.

File: rootfs/etc/entrypoint.d/60-apps.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickjahns) [reply rootfs/etc/entrypoint.d/60-apps.sh]: Can you elaborate why it makes sense to permanently uninstall an app on start of the container? It's somewhat of a one shot admin command. 

🤔 - also while being quite nice for testing - the APPS_INSTALL command is also a bit "overstepping boundaries"

[the remark this answers is not in the document]
Comment (tboerger) [reply rootfs/etc/entrypoint.d/60-apps.sh]: The APPS_INSTALL is exactly like that in use, so this is nothing I will remove. And I also had a request to remove some shipped app, so it MUST be executed on every container start.

File: rootfs/etc/entrypoint.d/85-others.sh [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+declare -x OWNCLOUD_ENTRYPOINT_INITIALIZED
```
Comment (patrickjahns) [inline rootfs/etc/entrypoint.d/85-others.sh]: shouldn't we run this as last in the order (i.e. 99? )

Comment (tboerger) [reply rootfs/etc/entrypoint.d/85-others.sh]: It can be even added to 00-foo.sh, that should not matter. It just sets the env variable to reduce double or tripple loading of the entrypoint scripts.

Comment (tboerger) [reply rootfs/etc/entrypoint.d/85-others.sh]: Besides that, there is already 99-apache.sh from the php base image.

**Comment by tboerger** [comment]
So far other ownCloud images are switched to `bionic-20181008` or `xenial-20181008` tags, so I will get to merge it now to get the new 10.0.10 image rolling.
